File: frontend/src/types/billing.ts
// Mirror of the backend billing_service.py responses.
// Kept loose (no discriminated unions) because the payload evolves as we
// tune the cost model — an over-tight type here would need churn on every
// tuning PR without adding safety.

export interface BillingTotals {
  messages: number;
  llm_usd: number;
  external_usd: number;
  total_usd: number;
}

export interface BillingProviderEntry {
  provider: string;
  total_usd: number;
}

export interface BillingChannelEntry {
  channel: string;
  messages: number;
  llm_usd: number;
  external_usd: number;
  total_usd: number;
}

export interface BillingDailyPoint {
  date: string;
  webchat: number;
  whatsapp: number;
  voice: number;
  total: number;
}

export interface OrgBillingRow {
  org_id: string;
  org_name: string;
  plan: string;
  primary_provider: string;
  messages: number;
  llm_usd: number;
  external_usd: number;
  total_usd: number;
  by_channel: Record<string, { messages: number; total_usd: number }>;
}

export interface PricingNotes {
  input_tokens_per_msg: number;
  output_tokens_per_msg: number;
  channel_extra_usd_per_msg: Record<string, number>;
  provider_model_map: Record<string, string>;
  model_input_price_per_1k: Record<string, number>;
  model_output_price_per_1k: Record<string, number>;
}

export interface BillingAnalyticsResponse {
  window_days: number;
  generated_at: string;
  org_id: string | null;
  totals: BillingTotals;
  by_provider: BillingProviderEntry[];
  by_channel: BillingChannelEntry[];
  daily: BillingDailyPoint[];
  per_org: OrgBillingRow[];
  pricing_notes: PricingNotes;
}

// ─── Org-staff snapshot ───────────────────────────────────────────

export interface PlanInfo {
  key: "free" | "starter" | "pro" | "enterprise" | string;
  display_name: string;
  monthly_message_quota: number;
  price_usd_monthly: number;
  features: string[];
}

export interface UsageInfo {
  messages_this_month: number;
  quota: number;
  percent_used: number;
}

export interface OrgBillingSnapshot {
  plan: PlanInfo;
  usage: UsageInfo;
  current_month_costs: BillingTotals;
  by_channel: BillingChannelEntry[];
  daily: BillingDailyPoint[];
  pricing_notes: PricingNotes;
}

File: backend/app/schemas/billing.py
import uuid
from datetime import datetime

from pydantic import BaseModel


# ─── Super-admin billing analytics response ──────────────────────────


class BillingKpis(BaseModel):
    total_ai_messages: int
    total_usd: float
    llm_input_usd: float
    llm_output_usd: float
    external_api_usd: float
    cost_per_message_usd: float


class BillingDailyPoint(BaseModel):
    date: str  # ISO YYYY-MM-DD
    total_usd: float


class BillingProviderEntry(BaseModel):
    provider: str
    total_usd: float


class BillingChannelEntry(BaseModel):
    channel: str
    total_usd: float


class OrgBillingRow(BaseModel):
    org_id: str
    org_name: str
    plan: str
    provider: str
    ai_messages: int
    llm_input_usd: float
    llm_output_usd: float
    external_api_usd: float
    total_usd: float
    channels: dict[str, float]


class BillingAnalyticsResponse(BaseModel):
    window_days: int
    generated_at: datetime
    org_id: str | None
    kpis: BillingKpis
    cost_by_day: list[BillingDailyPoint]
    cost_by_provider: list[BillingProviderEntry]
    cost_by_channel: list[BillingChannelEntry]
    per_org: list[OrgBillingRow]


# ─── Org-staff current-month billing snapshot ────────────────────────


class PlanInfo(BaseModel):
    key: str
    display_name: str
    monthly_message_quota: int
    price_usd_monthly: int
    features: list[str]


class UsageInfo(BaseModel):
    messages_used: int
    messages_quota: int
    percent_used: float
    ai_messages_this_month: int


class ProviderInfo(BaseModel):
    key: str
    model: str
    llm_input_price_per_1k: float
    llm_output_price_per_1k: float
    input_tokens_per_message: int
    output_tokens_per_message: int


class ChannelCostEntry(BaseModel):
    channel: str
    ai_messages: int
    total_usd: float


class CostThisMonth(BaseModel):
    llm_input_usd: float
    llm_output_usd: float
    external_api_usd: float
    total_usd: float
    by_channel: list[ChannelCostEntry]


class OrgBillingSnapshotResponse(BaseModel):
    org_id: uuid.UUID
    generated_at: datetime
    month_start: datetime
    plan: PlanInfo
    usage: UsageInfo
    provider: ProviderInfo
    channel_extras_per_msg: dict[str, float]
    cost_this_month: CostThisMonth
